Add Mock DDI timing, agonism, badness exportation

shawa opened the pull request:
There is now a shell script which will export a CSV file in the [reference format as previously given](https://github.com/andrewbutterfield/CS4098-2017/blob/master/ConstructDDItest.csv).

Either agonism or harmful can be selected as the third column.

shawa committed:
Use sha hashing, add agonism

shawa committed:
Merge branch 'iteration-7' into the-agony

shawa committed:
Add CSV exporting to `enrich.py`

shawa committed:
use sha1 over weirdo 224

shawa committed:
Add mock ddi exportation

shawa committed:
Merge branch 'master' into the-agony

houli commented on asclepius/asclepius/enrich.py:
```diff
@@ -1,8 +1,40 @@
+    writer.writeheader()
+
+    for ddi in enrich(dinto.all_ddis()):
+        ddi.update({'time': ddi['spacing'], 'unit': 'hr'})
```
Can we do random units too so not everything will be hours? Just for the sake of amusement of individuals

houli commented on export-mock-ddi-data.sh:
```diff
@@ -0,0 +1,60 @@
+            local readonly LOAD_PID="$!";
+            trap  "kill $LOAD_PID 2>&1 > /dev/null; exit" SIGTERM SIGKILL SIGINT
+            disown
+            docker-compose --file docker-compose.dev.yml \
```
He won't have the `latest` images just use the regular compose file which means we don't have to pass this option

houli commented on export-mock-ddi-data.sh:
```diff
@@ -0,0 +1,60 @@
+            trap  "kill $LOAD_PID 2>&1 > /dev/null; exit" SIGTERM SIGKILL SIGINT
+            disown
+            docker-compose --file docker-compose.dev.yml \
+                run --rm asclepius python3 asclepius/enrich.py "$flavour" \
```
The docker compose commands should also be ran with `sudo`
Also we should use `-p export` or something to namespace

houli commented on export-mock-ddi-data.sh:
```diff
@@ -0,0 +1,60 @@
+            disown
+            docker-compose --file docker-compose.dev.yml \
+                run --rm asclepius python3 asclepius/enrich.py "$flavour" \
+                > "$outfile"
```
After we finish we should `sudo docker-compose down -p export` etc.
https://github.com/tom-and-the-toothfairies/pathways/blob/master/doc/FEATURES_RELEASE_2.md#end-to-end-automated-testing is pretty much what we're doing

houli commented on export-mock-ddi-data.sh:
```diff
@@ -0,0 +1,60 @@
+            loading &
+            local readonly LOAD_PID="$!";
+            trap  "kill $LOAD_PID 2>&1 > /dev/null; exit" SIGTERM SIGKILL SIGINT
+            disown
```
I don't understand the trapping, disowning, kill $LOAD_PID, can you explain?

shawa replied:
The export takes 5-10 secs, which is a bit unacceptable to have no output at all; loading spinner is on its own process

`$!` is the PID of the last forked process, `disown` removes the proc from the shell's job list (so it can't be `fg`d) after the script is dead.

`trap` sets a callback (passed as a string) to the given list of signals; so it kills the spinner proc when it's ctr-c'd/sent a kill signal/the parent proc dies

houli commented on export-mock-ddi-data.sh:
```diff
@@ -0,0 +1,60 @@
+                run --rm asclepius python3 asclepius/enrich.py "$flavour" \
+                > "$outfile"
+            kill "$LOAD_PID";
+            exit 1
```
1 is an unsuccessful exit

c-brenn commented on asclepius/asclepius/enrich.py:
```diff
@@ -1,8 +1,40 @@
+from clize.parameters import one_of
+
+def _sha_int(string):
+    return int(sha1(string.encode()).hexdigest()[-8:], 16)
```
seeing as `int()` can take a base - could we skip the `sha1()` step and just use a larger base?

shawa replied:
As in, cast the bytes of the str to an int?

c-brenn replied:
That would probably work too. I was suggesting `int(string, base)` where `base` is like 256 (something big enough to account for all valid `uri` chars). Basically pretend the `uri` is some base 256 number.

Whatever you want to go with.

shawa replied:
Ah, wouldn't work; Python only supports parsing to 2 ≤ base ≤ 26 numbers. I don't think the sha stuff is much of a bottleneck; it's dwarfed by the time to query Chiron in any case.

shawa committed:
Add privilege escalation prompt, clarifiy PID name, fix exit statuses

shawa committed:
Drop shifts, clarify spinner PID, sudo everything, privilege escalator

shawa committed:
rm --rm

shawa committed:
Can't trap sigkill m80

shawa committed:
Merge branch 'the-agony' of github.com:tom-and-the-toothfairies/pathways into the-agony

houli commented on asclepius/asclepius/enrich.py:
```diff
@@ -1,8 +1,43 @@
+import clize
+from clize.parameters import one_of
+
+UNITS = ('hr', 'day')
```
I know it doesn't make much sense but can we do all of his units, (sec, min, hr, day, week, yr)?

c-brenn commented:
Looking good. Could you add this to the change log and document the new features in the features doc?

houli commented on asclepius/asclepius/enrich.py:
```diff
@@ -1,8 +1,43 @@
+
+
+def main(flavour:one_of('harmful', 'agonism')):
+    fieldnames = ('drug_a', 'drug_b', flavour, 'time', 'unit')
```
Can we use his field names "Drug 1, Drug 2, DDI Type, Time, Unit" with same capitalisation
where DDI Type values are "good"/"bad" and "agonistic"/"antagonistic" depending on the mode we're working in

houli commented on export-mock-ddi-data.sh:
```diff
@@ -0,0 +1,71 @@
+}
+
+if [ $EUID != 0 ]; then
+    sudo "$0" "$@"
```
do we need this if we're running sudo docker compose? it should prompt like the install docker script does

shawa replied:
Ah, that's to get the pwd prompt asap, before launching the spinner. Otherwise there'll be some load of slashes and pipes being printed as user types.

I feel like this spinner is adding much complexity. 🤔

shawa replied:
Realising now that running the script as root will also make the resultant CSV owned by root, which will cause much confusion. Will fix.

shawa committed:
Add units and more correc time intervals to mock DDIs

Include a block until chiron becomes available
Export data into mock-enriched-ddis dir for ease
Remove the spinner

shawa committed:
Update Features doc to include mock ddi data tests

houli committed:
Units are properly displayed in the front end

shawa committed:
Update CHANGELOG.md
